filebeat/tests/integration/filestream_test.go

```diff
@@ -657,6 +667,354 @@ func TestFilestreamTakeOverFromLogInput(t *testing.T) {
+
+			logFile := filepath.Join(workDir, "log.log")
+			// Escape filepaths for Windows
+			msgLogFilePath := logFile
+			if runtime.GOOS == "windows" {
+				msgLogFilePath = strings.ReplaceAll(logFile, `\`, `\\`)
+			}
+			integration.GenerateLogFile(t, logFile, 100, false)
+
+			vars := map[string]any{
+				"homePath":    workDir,
+				"logfile":     logFile,
+				"testdata":    testDataPath,
+				"esHost":      esAddr,
+				"gracePeriod": tc.gracePeriod.String(),
+			}
+			cfgYAML := getConfig(t, vars, "delete", tc.configTmpl)
+			filebeat.WriteConfigFile(cfgYAML)
+			filebeat.Start()
+
+			for _, msgFmt := range tc.msgs {
+				msg := fmt.Sprintf(msgFmt, msgLogFilePath)
+				filebeat.WaitForLogs(
+					msg,
+					10*time.Second,
+					"did not find '%s' in the logs",
+					msg,
+				)
+			}
+
+			if tc.resourceNotFinished {
+				// Wait a few times for the 'not finished' logs
+				notFinishedMsg := fmt.Sprintf(
+					"not all events from '%s' have been published, "+
+						"closing harvester",
+					msgLogFilePath)
+				for i := range 2 {
+					filebeat.WaitForLogs(
+						notFinishedMsg,
+						10*time.Second,
+						"[%d] Filebeat did not wait for the resource to be finished",
+						i,
+					)
+				}
+
+				if tc.dataAdded {
+					// Add more data to the log file
+					integration.GenerateLogFile(t, msgLogFilePath, 5, true)
+
+					// Wait for the "file update" log
+					filebeat.WaitForLogs(
+						fmt.Sprintf("File %s has been updated", msgLogFilePath),
+						time.Second,
+						"filewatcher did not detect the file as updated")
+
+					// Wait for the harvester to be closed
+					filebeat.WaitForLogs(
+						fmt.Sprintf("not all events from '%s' have been published, closing harvester", msgLogFilePath),
+						10*time.Second,
+						"harvester was not closed after data added to the file")
+
+					// Wait for the "not changed" log
+					filebeat.WaitForLogs(
+						fmt.Sprintf("File %s has not changed, trying to start new harvester because 'delete' is enabled", msgLogFilePath),
+						time.Second,
+						"Filestream did not try to start a new harvester for the unchanged file")
+
+					// Ensure harvester closes without removing the file
+					filebeat.WaitForLogs(
+						fmt.Sprintf("not all events from '%s' have been published, closing harvester", msgLogFilePath),
+						10*time.Second,
+						"Harvester was not closed because the resource is not finished")
+
+					if !fileExists(t, logFile) {
+						t.Fatalf("%q should not have been removed", logFile)
+					}
+				}
+
+				if err := es.UpdateOdds(0, 0, 0, 0); err != nil {
+					t.Fatalf("cannot update mock-es odds: %s", err)
+				}
+			}
+
+			msg := fmt.Sprintf("'%s' removed", msgLogFilePath)
+			filebeat.WaitForLogs(msg, 30*time.Second, "file removed log entry not found")
+			removedMsg := filebeat.GetLastLogLine(msg)
+
+			gracePeriodMsg := fmt.Sprintf("all events from '%s' have been published, waiting for %s grace period", msgLogFilePath, tc.gracePeriod)
```

Comment:
You are validating that Filebeat thinks it shipped the events, can you validate that they actually were shipped by querying them out of ES or using the file output to read them back?

I see why you are using log statements, just keep in mind that testing this instead of always looking at the externally observable effects can cause you to miss bugs. 

If any of these logs messages were generated incorrectly because of a bug, your test could still believe the feature did the correct thing.

Reply:
I'll add the check for the events published on ES.

Reply:
Actually, I'm not using a real ES on those tests. I'm using [`mock-es`](https://github.com/elastic/mock-es), the maximum visibility I have is the number of items accepted by the bulk API.

Would that be enough for you?

Reply:
It depends, does that number increment for retries of the same event or is it for unique documents accepted? If it is the latter it will confirm you shipped everything, if it is the first one it won't.